File: 백준/Silver/9012. 괄호/괄호.c
#include <stdio.h>

char stack[50];
int top = 0;

int is_empty() {
	return top == 0;
}
void push(char c) {
	stack[top++] = c;
}

char pop() {
	return stack[--top];
}

int is_VPS(char* str) {
	int result = 1;

	for (int i = 0; str[i]; i++) {
		if (str[i] == '(')
			push(str[i]);
		else {
			if (is_empty()) {
				result = 0;
				break;
			}
			else pop();
		}
	}
	if (!is_empty()) result = 0;
	while (!is_empty())
		pop();

	return result;
}

int main(void) {
	int t;
	char str[50];

	scanf("%d", &t); 

	for (int i = 0; i < t; i++) {
		scanf("%s", str);

		if (is_VPS(str)) printf("YES \n");
		else printf("NO \n");
	}

	return 0;
}

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Stack;

public class Main{
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder sb = new StringBuilder();
        int n = Integer.parseInt(br.readLine());

        for(int i = 0; i < n; i++){
            String line = br.readLine();
            if(isVps(line))
                sb.append("YES\n");
            else
                sb.append("NO\n");
        }

        bw.append(sb.toString());
        bw.flush();
        bw.close();
    }

    public static boolean isVps(String s){
        Stack<Character> stack = new Stack<>();

        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);

            if(c == '(')
                stack.push(c);
            else if(c == ')'){ //짝이 있을 경우 pop
                if(stack.isEmpty())
                    return false;
                else stack.pop();
            } 
        }
        return stack.isEmpty(); //비었는지 확인
    }
}
